tools/tool-testing/selftest.js

```diff
@@ -1856,8 +1868,108 @@ class TestList {
   // Mark a test's file as having failures.  This prevents
   // saveTestState from saving its hash as a potentially
   // "unchanged" file to be skipped in a future run.
-  notifyFailed(test) {
+  notifyFailed(test, failure) {
     this.fileInfo[test.file].hasFailures = true;
+    test.failure = failure || true;
```

Comment:
I think the signature could be better for this.  Perhaps `notifyFailed(test, failureObject = null)`, and it could set a variable called `test.failed = true` (always) and `test.failureObject = failureObject`.  I'll adjust tomorrow unless there is an objection.